File: VB/VB/DatabaseUserSettings/Controllers/ManageUserSettingsWindowController.vb
Imports DevExpress.ExpressApp
Imports DevExpress.Data.Filtering
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Utils
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.ExpressApp.Security

Namespace DatabaseUserSettings.Controllers
    Public Class ManageUserSettingsWindowController
        Inherits WindowController

        Public Const CriteriaKeyAllUsersExceptConfigurator As String = "AllUsersExceptConfigurator"
        Public Const ActiveKeyConfiguratorOnly As String = "ConfiguratorOnly"
        Private Const ActiveKeyAlwaysInactive As String = "AlwaysInactive"
        Private ReadOnly _manageUserSettingsAction As PopupWindowShowAction
        Private [module] As DatabaseUserSettingsModule
        Public Sub New()
            TargetWindowType = WindowType.Main
            _manageUserSettingsAction = New PopupWindowShowAction(Me, "ManageUserSettings", PredefinedCategory.Tools)
            AddHandler _manageUserSettingsAction.CustomizePopupWindowParams, AddressOf ManageUserSettingsAction_CustomizePopupWindowParams
            _manageUserSettingsAction.ImageName = "BO_Department"
            _manageUserSettingsAction.ToolTip = "Allows you to manage user settings for all application users."
        End Sub
        Protected ReadOnly Property DbModelStoreModule() As DatabaseUserSettingsModule
            Get
                Return [module]
            End Get
        End Property
        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            Guard.ArgumentNotNull(Application, "application")
            [module] = CType(Application.Modules.FirstOrDefault(Function(m) m.GetType() Is GetType(DatabaseUserSettingsModule)), DatabaseUserSettingsModule)
            Guard.ArgumentNotNull([module], "module")
            UpdateManageSettingsActions()
            UpdateUserSettings()
        End Sub
        Protected Overridable Sub UpdateManageSettingsActions()
            Using objectSpace As IObjectSpace = Application.CreateObjectSpace()
                _manageUserSettingsAction.Active(ActiveKeyConfiguratorOnly) = SecuritySystem.CurrentUserName = DatabaseUserSettingsModule.ConfiguratorUserName OrElse (SecuritySystem.CurrentUser IsNot Nothing AndAlso objectSpace.IsObjectFitForCriteria(SecuritySystem.CurrentUser, New FunctionOperator(IsCurrentUserInRoleOperator.OperatorName, New OperandValue(DatabaseUserSettingsModule.ConfiguratorRoleName))).GetValueOrDefault(False))
            End Using
        End Sub
        Protected Overridable Sub UpdateUserSettings()
            Using objectSpace As IObjectSpace = Application.CreateObjectSpace()
                If SecuritySystem.UserType IsNot Nothing Then
                    For Each user As Object In objectSpace.GetObjects(SecuritySystem.UserType)
                        DatabaseUserSettingsStore.GetUserSettings(objectSpace, DbModelStoreModule.UserSettingsType, DatabaseUserSettingsModule.UserIdTypeConverter.ConvertToInvariantString(objectSpace.GetKeyValue(user)))
                    Next user
                End If
                objectSpace.CommitChanges()
            End Using
        End Sub
        Private Sub ManageUserSettingsAction_CustomizePopupWindowParams(ByVal sender As Object, ByVal e As CustomizePopupWindowParamsEventArgs)
            SetupPopupWindow(e)
        End Sub
        Protected Overridable Sub SetupPopupWindow(ByVal e As CustomizePopupWindowParamsEventArgs)
            Dim objectSpace As IObjectSpace = e.Application.CreateObjectSpace()
            Dim detailView As DetailView = e.Application.CreateDetailView(objectSpace, objectSpace.CreateObject(DbModelStoreModule.ManageUserSettingsParameterType), False)
            detailView.ViewEditMode = ViewEditMode.Edit
            e.DialogController.AcceptAction.Active(ActiveKeyAlwaysInactive) = False
            e.View = detailView
        End Sub
        Public ReadOnly Property ManageUserSettingsAction() As PopupWindowShowAction
            Get
                Return _manageUserSettingsAction
            End Get
        End Property
    End Class
End Namespace

File: CS/CS/DatabaseUserSettings/Controllers/ManageUserSettingsWindowController.cs
using System;
using System.Linq;
using DevExpress.ExpressApp;
using DevExpress.Data.Filtering;
using System.Collections.Generic;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp.Utils;
using DevExpress.ExpressApp.Editors;
using DevExpress.ExpressApp.Actions;
using DevExpress.ExpressApp.Security;

namespace DatabaseUserSettings.Controllers {
    public class ManageUserSettingsWindowController : WindowController {
        public const string CriteriaKeyAllUsersExceptConfigurator = "AllUsersExceptConfigurator";
        public const string ActiveKeyConfiguratorOnly = "ConfiguratorOnly";
        private const string ActiveKeyAlwaysInactive = "AlwaysInactive";
        private readonly PopupWindowShowAction manageUserSettingsAction;
        private DatabaseUserSettingsModule module;
        public ManageUserSettingsWindowController() {
            TargetWindowType = WindowType.Main;
            manageUserSettingsAction = new PopupWindowShowAction(this, "ManageUserSettings", PredefinedCategory.Tools);
            manageUserSettingsAction.CustomizePopupWindowParams += ManageUserSettingsAction_CustomizePopupWindowParams;
            manageUserSettingsAction.ImageName = "BO_Department";
            manageUserSettingsAction.ToolTip = "Allows you to manage user settings for all application users.";
        }
        protected DatabaseUserSettingsModule DbModelStoreModule { get { return module; } }
        protected override void OnActivated() {
            base.OnActivated();
            Guard.ArgumentNotNull(Application, "application");
            module = (DatabaseUserSettingsModule)Application.Modules.FirstOrDefault<ModuleBase>(
                m => m.GetType() == typeof(DatabaseUserSettingsModule)
            );
            Guard.ArgumentNotNull(module, "module");
            UpdateManageSettingsActions();
            UpdateUserSettings();
        }
        protected virtual void UpdateManageSettingsActions() {
            using (IObjectSpace objectSpace = Application.CreateObjectSpace()) {
                manageUserSettingsAction.Active[ActiveKeyConfiguratorOnly] = SecuritySystem.CurrentUserName == DatabaseUserSettingsModule.ConfiguratorUserName
                    || (SecuritySystem.CurrentUser != null && objectSpace.IsObjectFitForCriteria(
                        SecuritySystem.CurrentUser,
                        new FunctionOperator(IsCurrentUserInRoleOperator.OperatorName, new OperandValue(DatabaseUserSettingsModule.ConfiguratorRoleName))
                    ).GetValueOrDefault(false)
                );
            }
        }
        protected virtual void UpdateUserSettings() {
            using (IObjectSpace objectSpace = Application.CreateObjectSpace()) {
                if (SecuritySystem.UserType != null)
                    foreach (object user in objectSpace.GetObjects(SecuritySystem.UserType))
                        DatabaseUserSettingsStore.GetUserSettings(objectSpace, DbModelStoreModule.UserSettingsType, DatabaseUserSettingsModule.UserIdTypeConverter.ConvertToInvariantString(objectSpace.GetKeyValue(user)));
                objectSpace.CommitChanges();
            }
        }
        private void ManageUserSettingsAction_CustomizePopupWindowParams(object sender, CustomizePopupWindowParamsEventArgs e) {
            SetupPopupWindow(e);
        }
        protected virtual void SetupPopupWindow(CustomizePopupWindowParamsEventArgs e) {
            IObjectSpace objectSpace = e.Application.CreateObjectSpace();
            DetailView detailView = e.Application.CreateDetailView(objectSpace, objectSpace.CreateObject(DbModelStoreModule.ManageUserSettingsParameterType), false);
            detailView.ViewEditMode = ViewEditMode.Edit;
            e.DialogController.AcceptAction.Active[ActiveKeyAlwaysInactive] = false;
            e.View = detailView;
        }
        public PopupWindowShowAction ManageUserSettingsAction { get { return manageUserSettingsAction; } }
    }
}